test/lit/passes/type-merging.wast

```diff
@@ -0,0 +1,198 @@
+  ;; CHECK-NEXT:  (local $d (ref null $D))
+  ;; CHECK-NEXT:  (local $e (ref null $D))
+  ;; CHECK-NEXT:  (local $f (ref null $D))
+  ;; CHECK-NEXT:  (local $g (ref null $D))
+  ;; CHECK-NEXT:  (nop)
+  ;; CHECK-NEXT: )
+  (func $foo
+    (local $a (ref null $A))
+    ;; $B can be merged into $A.
+    (local $b (ref null $B))
+    ;; $C can be merged into $B, so it will merge into $A.
+    (local $c (ref null $C))
+    ;; $D cannot be merged into $A as it adds a field.
+    (local $d (ref null $D))
+    ;; $E can be merged into $D.
+    (local $e (ref null $E))
+    ;; $F can be merged into $E, so it will merge into $D.
+    (local $f (ref null $F))
+    ;; $G can be merged into $F, so it will merge into $D.
+    (local $g (ref null $G))
+  )
+)
+
+(module
+  ;; CHECK:      (type $A (struct (field (ref null $A))))
+  (type $A (struct_subtype (field (ref null $A)) data))
+  (type $B (struct_subtype (field (ref null $A)) $A))
+  ;; CHECK:      (type $none_=>_none (func))
+
+  ;; CHECK:      (type $C (struct_subtype (field (ref null $A)) $A))
+  (type $C (struct_subtype (field (ref null $B)) $A))
+
+  ;; CHECK:      (func $foo (type $none_=>_none)
+  ;; CHECK-NEXT:  (local $a (ref null $A))
+  ;; CHECK-NEXT:  (local $b (ref null $A))
+  ;; CHECK-NEXT:  (local $c (ref null $C))
+  ;; CHECK-NEXT:  (nop)
+  ;; CHECK-NEXT: )
+  (func $foo
+    ;; $A will remain the same.
+    (local $a (ref null $A))
+    ;; $B can be merged into $A.
+    (local $b (ref null $B))
+    ;; $C refines the field, so it cannot be merged. However, separately, in
+    ;; the type definition of $C, its field of type $B should become $A. That
+    ;; is, $B should no longer be used anywhere.
+    (local $c (ref null $C))
+  )
+)
+
+;; Check that we refinalize properly.
+(module
+  ;; CHECK:      (type $A (struct ))
+  (type $A (struct))
+  (type $B (struct_subtype $A))
+
+  ;; CHECK:      (type $none_=>_ref?|$A| (func (result (ref null $A))))
+
+  ;; CHECK:      (func $returner (type $none_=>_ref?|$A|) (result (ref null $A))
+  ;; CHECK-NEXT:  (local $local (ref null $A))
+  ;; CHECK-NEXT:  (local.get $local)
+  ;; CHECK-NEXT: )
+  (func $returner (result (ref null $B))
+    (local $local (ref null $B))
+
+    ;; After we change the local to use type $A, we need to update the local.get's
+    ;; type as well, or we will error.
+    (local.get $local)
+  )
+)
+
+;; Test some real-world patterns, including fields to ignore, links between
+;; merged types, etc.
+;;
+;; The result here is that we will merge $type$3$to-merge into $type$3, and
+;; $type$6$to-merge into $type$6. While doing so we must update the fields and
+;; the expressions that they appear in, and not error.
+(module
+  ;; CHECK:      (type $type$5 (struct (field (mut i32))))
+
+  ;; CHECK:      (type $type$6 (struct_subtype (field (mut i32)) (field (mut i32)) $type$5))
+
+  ;; CHECK:      (type $type$2 (array (mut (ref null $type$5))))
+  (type $type$2 (array (mut (ref null $type$5))))
```

Comment:
This test would be easier to read if it used shorter and more descriptive names, like `$A` and `$B` in the previous tests. The `to-merge` suffix is helpful, though.

Reply:
Good idea, done.